lib/src/Pos/Util/UserSecret.hs

```diff
@@ -71,51 +90,68 @@ import           System.Wlog (logWarning)
+    , _adAddresses :: !(Vector (Word32, Address))
+    -- ^ First value is path, second value is address which can be
+    -- computed from root key and path, but it would require a
+    -- passphrase.
     }
 
-makeLenses ''WalletUserSecret
+makeLenses ''AccountData
 
-instance Buildable WalletUserSecret where
-    build WalletUserSecret{..} =
-        bprint ("{ root = "%addressF%", set name = "%build%
-                ", wallets = "%pairsF%", accounts = "%pairsF%" }")
-        (makeRootPubKeyAddress $ encToPublic _wusRootKey)
-        _wusWalletName
-        _wusAccounts
-        _wusAddrs
+deriveSimpleBi ''AccountData [
+    Cons 'AccountData [
+        Field [| _adName      :: Text                     |],
+        Field [| _adPath      :: Word32                   |],
+        Field [| _adAddresses :: Vector (Word32, Address) |]
+    ]]
+
+instance Buildable AccountData where
+    build AccountData {..} =
+        bprint ("{ name = "%stext%", path = ["%int%"]"%
+                ", addresses = "%pairsF%" }")
+        _adName _adPath _adAddresses
       where
-        pairsF :: (Buildable a, Buildable b) => Format r ([(a, b)] -> r)
-        pairsF = later $ mconcat . map (uncurry $ bprint ("("%build%", "%build%")"))
+        pairsF :: (Buildable a, Buildable b) => Format r (Vector (a, b) -> r)
+        pairsF = later $ fold . map (uncurry $ bprint ("("%build%", "%build%")"))
 
-deriveSimpleBi ''WalletUserSecret [
-    Cons 'WalletUserSecret [
-        Field [| _wusRootKey    :: EncryptedSecretKey |],
-        Field [| _wusWalletName :: Text               |],
-        Field [| _wusAccounts   :: [(Word32, Text)]   |],
-        Field [| _wusAddrs      :: [(Word32, Word32)] |]
+data WalletData = WalletData
+    { _wdRootKey  :: !EncryptedSecretKey
+    , _wdName     :: !Text
+    , _wdAccounts :: !(Vector AccountData)
+    }
+
+makeLenses ''WalletData
+
+deriveSimpleBi ''WalletData [
+    Cons 'WalletData [
+        Field [| _wdRootKey    :: EncryptedSecretKey |],
+        Field [| _wdName       :: Text               |],
+        Field [| _wdAccounts   :: Vector AccountData |]
     ]]
 
-mkGenesisWalletUserSecret :: EncryptedSecretKey -> WalletUserSecret
-mkGenesisWalletUserSecret _wusRootKey = do
-    let _wusWalletName = "Genesis wallet"
-        _wusAccounts   = [(accountGenesisIndex, "Genesis account")]
-        _wusAddrs      = [(accountGenesisIndex, wAddressGenesisIndex)]
-    WalletUserSecret{..}
+instance Buildable WalletData where
+    build WalletData {..} =
+        bprint ("{ name = "%stext%", accounts = "%listJson%" }")
+        _wdName _wdAccounts
 
+----------------------------------------------------------------------------
+-- UserSecret
+----------------------------------------------------------------------------
 
 -- | User secret data. Includes secret keys only for now (not
 -- including auxiliary @_usPath@).
 data UserSecret = UserSecret
     { _usKeys    :: [EncryptedSecretKey]
```

Comment:
These keys are already in `_usWallets`. Please remove duplicated data.
(I understand that `_usKeys` is possibly used somewhere in the Core, but I believe that there are not too many use cases and it's not very time consuming to rewrite them using `usWallets`)

Reply:
Okay.